backend/game/shop/shop.py

```diff
@@ -0,0 +1,63 @@
+import json
+import random
+
+from typing import List, Dict, TextIO, Tuple
+from math import exp, floor
+
+class Shop:
+
+    path_to_units_file: str = './../data/units.json'
+
+    def __init__(self):
+        with open(self.path_to_units_file, 'r') as units_f:
+            self.units_list = self.read_json(units_f)
+        self.units_amount = len(self.units_list)
+        self.add_prices()
+
+    def read_json(self, file: TextIO) -> List[Dict]:
+        return json.load(file)
+
+    def add_prices(self):
+        for unit in self.units_list:
+            price = self.calculate_price(unit)
+
+    def calculate_price(self, unit: Dict) -> int:
+        stats_sum = 0
+        for stat in unit['statistics']:
+            stats_sum += unit['statistics'][stat]
+        stats_sum -= unit['statistics']['base_hp']
+        value = self.func(stats_sum)
+        if value > 10:
+            value = 13
+
+        unit['price'] = value
+
+    def func(self, x: int) -> int:
+        # We need to think about costs, how much currency player is able to obtain?
+        return int(round((-0.01729982 + 0.3214279 * exp(0.09289161 * x))))
+
+    def get_random_units(self, quantity: int, currency: int) -> List[Dict]:
+        # If player is too poor to afford anything, return total random
+        if currency < 2:
+            return random.sample(self.units_list, quantity)
+        else:
+            # Else random few random units and few affordable units
+            units_to_return = []
+            random_amount = int(floor(quantity * 2 / 3))
+            units_to_return += random.sample(self.units_list, random_amount)
+            units_to_return += self.get_affordable_units(currency, (quantity - random_amount))
+            return units_to_return
+
+    def get_affordable_units(self, currency: int, quantity: int) -> List:
+        affordable_units = []
```

Comment:
return random.choices([u for u in self units_list if u['price'] <= currency], k=quantity)

Zamiast całej funkcji, jeśli chcemy jednostki z powtórzeniami tak jak jest teraz. Jak nie, to random.sample ale trzeba sprawdzić czy quantity nie jest za duże :smile:

Reply:
<3 dzięki wielkie, zmienię, myślałem chwilę jak to "spakować" w mniejszą linię, ale nie wpadłem.